### PR TITLE
Phase 1 adding user, articles and CRUD operations

### PR DESCRIPTION
**Pull request by bv-pulkit**

### CODE REVIEW

File: app/controllers/articles_controller.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
+class ArticlesController < ApplicationController
+	def index
+		@articles = Article.all
+	end
+
+	def show
+		@user = User.find(params[:user_id])
```
Comment (bv-hemant) [inline app/controllers/articles_controller.rb]: use before action and what if user is not present with that user id

File: app/controllers/articles_controller.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+		@article = @user.articles.find(params[:id])
```
Comment (bv-hemant) [inline app/controllers/articles_controller.rb]: what if user is nil

File: app/controllers/articles_controller.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+		if @article.save
+			redirect_to user_path(@user)
+		else
+			render :new, status: :unprocessable_entity
+		end
+	end
+
+	def edit
+		@user = User.find(params[:user_id])
+		@article = @user.articles.find(params[:id])
```
Comment (bv-hemant) [inline app/controllers/articles_controller.rb]: if there is no article with params[:id] ?

File: app/controllers/articles_controller.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	end
+
+	def new
+		@article = Article.new
+	end
+
+	def create
+		@user = User.find(params[:user_id])
+		@article = @user.articles.new(article_params)
+
```
Comment (bv-hemant) [inline app/controllers/articles_controller.rb]: remove extra line

File: app/controllers/articles_controller.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+		if @article.save
```
Comment (bv-hemant) [inline app/controllers/articles_controller.rb]: @article.save!

File: app/controllers/articles_controller.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+	def edit
+		@user = User.find(params[:user_id])
+		@article = @user.articles.find(params[:id])
+	end
+
+	def update
+		@user = User.find(params[:user_id])
+		@article = @user.articles.find(params[:id])
+
+		if @article.update(article_params)
```
Comment (bv-hemant) [inline app/controllers/articles_controller.rb]: @article.update!(article_params)

File: app/controllers/articles_controller.rb [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+		else
+			render :edit, status: :unprocessable_entity
+		end
+	end
+
+	def destroy
+		@user = User.find(params[:user_id])
+		@article = @user.articles.find(params[:id])
+		puts "ready to destroy"
+		@article.destroy
```
Comment (bv-hemant) [inline app/controllers/articles_controller.rb]: @article.destroy!

File: app/controllers/articles_controller.rb [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
```
Comment (bv-hemant) [inline app/controllers/articles_controller.rb]: remove extra line

File: app/controllers/articles_controller.rb [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	def edit
+		@user = User.find(params[:user_id])
+		@article = @user.articles.find(params[:id])
+	end
+
+	def update
+		@user = User.find(params[:user_id])
+		@article = @user.articles.find(params[:id])
+
```
Comment (bv-hemant) [inline app/controllers/articles_controller.rb]: extra line

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+		@user = User.new(user_params)
+		if @user.save
+			redirect_to root_path, notice: "Successfully created account"
+		else
+			render :new, status: :unprocessable_entity
+		end
+	end
+
+	def show
+		@user = User.find(params[:id])
```
Comment (bv-hemant) [inline app/controllers/users_controller.rb]: ignore find use where

File: app/views/articles/_form.html.erb [hunk 11]
```diff
@@ -0,0 +1,14 @@
+<%= form_with model: [@user, @user.articles.find(params[:id])] do |form| %>
```
Comment (bv-hemant) [inline app/views/articles/_form.html.erb]: why doing query if you already have @article

File: app/views/articles/show.html.erb [hunk 12]
```diff
@@ -0,0 +1,19 @@
+<h1>Article</h1>
+<p>
+<strong>Title:</strong><br>
+<%= @article.title %>
```
Comment (bv-hemant) [inline app/views/articles/show.html.erb:5]: indentation

File: db/migrate/20220817091234_create_articles.rb [hunk 13]
```diff
@@ -0,0 +1,11 @@
+class CreateArticles < ActiveRecord::Migration[7.0]
+  def change
+    create_table :articles do |t|
+      t.string :title
+      t.text :body
+      t.references :user, null: false, foreign_key: true
```
Comment (bv-hemant) [inline db/migrate/20220817091234_create_articles.rb:6]: t.integer :user_id

File: db/migrate/20220817091234_create_articles.rb [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+
```
Comment (bv-hemant) [inline db/migrate/20220817091234_create_articles.rb:7]: remove extra line

File: db/migrate/20220817091234_create_articles.rb [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+      t.timestamps
+    end
```
Comment (bv-hemant) [inline db/migrate/20220817091234_create_articles.rb:9]: add_index :articles, :user_id

File: db/migrate/20220817055448_create_users.rb [hunk 16]
```diff
@@ -0,0 +1,10 @@
+class CreateUsers < ActiveRecord::Migration[7.0]
+  def change
+    create_table :users do |t|
+      t.string :email, null: false
+      t.string :password_digest
+
```
Comment (bv-hemant) [inline db/migrate/20220817055448_create_users.rb]: extra line

**Review by bv-hemant — CHANGES_REQUESTED** [review]

File: app/controllers/articles_controller.rb [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+	before_action :load_user
+	before_action :load_article, except: [:index, :new, :create]
+
+	def index
+		@articles = @user.articles
+	end
+
+	def show
+	end
+
```
Comment (bv-harsh) [inline app/controllers/articles_controller.rb]: Check if this is needed.

File: app/controllers/articles_controller.rb [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+	def new
+		@article = Article.new
+	end
+
+	def create
+		@article = @user.articles.new(article_params)
```
Comment (bv-harsh) [inline app/controllers/articles_controller.rb]: extract the params and than create, save the article.

File: app/controllers/articles_controller.rb [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+		@user = User.where(:id => params[:user_id]).last
+		user_not_exists if !@user.present?
+	end
+
+	def load_article
+		@article = @user.articles.where(:id => params[:id]).last if @user
+		article_not_exists if !@article.present?
+	end
+
+	def article_not_exists
```
Comment (bv-harsh) [inline app/controllers/articles_controller.rb]: article_not_found

File: app/controllers/users_controller.rb [hunk 20]
```diff
@@ -0,0 +1,33 @@
+class UsersController < ApplicationController
+	def index
+		@users = User.all
+		user_exist_check if !@users.present?
```
Comment (bv-harsh) [inline app/controllers/users_controller.rb]: Already checked, concluding user_not_found

File: app/controllers/users_controller.rb [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+	end
+
+	def new
+		@user = User.new
+	end
+
+	def create
+		@user = User.new(user_params)
+		if @user.save
```
Comment (bv-harsh) [inline app/controllers/users_controller.rb:14]: @user.save!

File: app/controllers/users_controller.rb [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+		user_exist_check(@user)
+	end
+
+	private
+	def user_params
+		params.require(:user).permit(:email, :password, :password_confirmation)
+	end
+
+	def user_exist_check(user)
+      redirect_to new_user_path, notice: "User does not exist. Please Sign Up" if user.nil?
```
Comment (bv-harsh) [inline app/controllers/users_controller.rb]: indentation, Also improve this function

File: app/models/article.rb [hunk 23]
```diff
@@ -0,0 +1,6 @@
+class Article < ApplicationRecord
+	belongs_to :user
+
```
Comment (bv-harsh) [inline app/models/article.rb]: no need for extra line

File: app/models/user.rb [hunk 24]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+	has_many :articles
+	has_secure_password
+
+	validates :email, presence: true, format: {with: /\A[^@\s]+@[^@\s]+\z/, message: "must be a valid Email Address"}
```
Comment (bv-harsh) [inline app/models/user.rb]: break line, check once if there are any better validations present for email - https://stackoverflow.com/questions/38611405/email-validation-in-ruby-on-rails

File: app/views/articles/_form.html.erb (hunk 11 above)
Comment (bv-harsh) [reply app/views/articles/_form.html.erb]: Is this resolved?

Comment (bv-harsh) [reply app/views/articles/_form.html.erb]: Also find might return nil, can that cause any issue here?

File: app/views/articles/index.html.erb [hunk 25]
```diff
@@ -0,0 +1,5 @@
+<h1>Article</h1>
```
Comment (bv-harsh) [inline app/views/articles/index.html.erb]: Articles, I guess plural

File: app/views/articles/show.html.erb [hunk 26]
```diff
@@ -0,0 +1,20 @@
+<h1>Article</h1>
+<h2>Article By - <%= @user.email %> </h2>
+<p>
+	<strong>Title:</strong><br>
+	<%= @article.title %>
+</p>
```
Comment (bv-harsh) [inline app/views/articles/show.html.erb]: Here you are following no convention. After <p> tag somewhere you are putting tabs and sometimes you are not. We usually don't.

File: app/views/articles/show.html.erb [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+
+<ul>
+	<li><%= link_to "Edit", edit_user_article_path(@user, @article) %></li>
+	<li><%= button_to "Destroy", user_article_path(@user, @article), 
+		method: :delete,
+		form: {
+		data: {
+		turbo_confirm: "Are you sure?" }
+	}%></li>
```
Comment (bv-harsh) [inline app/views/articles/show.html.erb]:
```
method: :delete,
        form: {
                data: { turbo_confirm: "Are you sure?" }
        } %>
</li>
```

File: app/views/users/index.html.erb [hunk 28]
```diff
@@ -0,0 +1,8 @@
+<h1>This is the home page</h1>
```
Comment (bv-harsh) [inline app/views/users/index.html.erb]: Please improve

File: app/views/users/index.html.erb [hunk 29]
```diff
@@ -0,0 +1,8 @@
...
+<ul>
+	<% @users.each do |user| %>
+		<li>
+			<%= link_to user.email, user %>
```
Comment (bv-harsh) [inline app/views/users/index.html.erb:6]: is this working?
Please also add screenshots of all the pages.

**Review by bv-harsh — CHANGES_REQUESTED** [review]

File: app/controllers/articles_controller.rb (hunk 17 above)
Comment (bv-pulkit) [reply app/controllers/articles_controller.rb]: Not needed

"Rails automatically infers the view from action names in routes.
If you're not defining data (only to be done in the controller), you'll can rely on the view being loaded without the action being present:"

File: app/views/users/index.html.erb (hunk 29 above)
Comment (bv-pulkit) [reply app/views/users/index.html.erb:6]: yes its working. will add screenshots

Comment (bv-pulkit) [reply app/views/users/index.html.erb:6]: image

**Review by bv-hemant — APPROVED** [review]

**Pull request merged**